### PR TITLE
Fix Psalm-issues batch #2

### PR DESCRIPTION
**Pull request by tvdijen**
Fixes all Psalm-issues in all subdirs under `modules/`.

### CODE REVIEW

File: modules/core/www/idp/logout-iframe.php [hunk 1]
```diff
@@ -115,9 +115,9 @@
 
 $globalConfig = \SimpleSAML\Configuration::getInstance();
 if ($type === 'nojs') {
-    $t = new \SimpleSAML\XHTML\Template($globalConfig, 'core:logout-iframe-wrapper.php');
+    $t = new \SimpleSAML\XHTML\Template($globalConfig, 'core:logout-iframe-wrapper.tpl.php');
```
Comment (tvdijen) [inline modules/core/www/idp/logout-iframe.php]: By giving the old PHP-templates a .tpl.php extension, they can be easily ignored by Psalm.

File: modules/core/lib/Auth/Process/AttributeAlter.php [hunk 2]
```diff
@@ -175,13 +175,15 @@ public function process(&$request)
...
                 );
             } else {
+                /** @var array $attr */
+                $attr = $attributes[$this->subject];
                 $attributes[$this->target] = array_diff(
                     preg_replace(
                         $this->pattern,
                         $this->replacement,
-                        $attributes[$this->subject]
+                        $attr
```
Comment (tvdijen) [inline modules/core/lib/Auth/Process/AttributeAlter.php]: Psalm is smart enough to know that if arg3 is an array, the return type of preg_replace is an array as well..
It is however not smart enough to know that $attributes[$this->subject] is an array, so we have to annotate it.. Annotation inside the preg_replace doesn't work, so I had to do it earlier.

File: modules/saml/lib/IdP/SAML2.php [hunk 3]
```diff
@@ -793,6 +809,7 @@ public static function getHostedMetadata($entityid)
             $hasNewCert = true;
         }
 
+        /** @var array $certInfo */
```
Comment (tvdijen) [inline modules/saml/lib/IdP/SAML2.php]: Second arg = true, means it will always return an array

File: modules/saml/lib/Message.php [hunk 4]
```diff
@@ -415,7 +417,11 @@ private static function decryptAssertion(
             }
         }
 
-        /** @var \Exception $lastException */
+        /**
+         * The annotation below is not working - See vimeo/psalm#1909
+         * @psalm-suppress InvalidThrow
```
Comment (tvdijen) [inline modules/saml/lib/Message.php]: Posted it as a feature request to support annotation on throw-statements.

File: modules/saml/www/idp/certs.php [hunk 5]
```diff
@@ -18,19 +18,22 @@
 
 switch ($_SERVER['PATH_INFO']) {
     case '/new_idp.crt':
-        $certInfo = SimpleSAML\Utils\Crypto::loadPublicKey($idpmeta, false, 'new_');
```
Comment (tvdijen) [inline modules/saml/www/idp/certs.php]: Doesn't make sense to _not_ require its existence and then display an undefined index error below.
Set false > true to require it to exist. That will also guarantee us it returns an array

File: modules/saml/www/sp/metadata.php [hunk 6]
```diff
@@ -208,9 +208,11 @@
 // add technical contact
 $email = $config->getString('technicalcontact_email', 'na@example.org');
 if ($email && $email !== 'na@example.org') {
-    $techcontact['emailAddress'] = $email;
```
Comment (tvdijen) [inline modules/saml/www/sp/metadata.php]: Psalm doesn't like this because $techcontact is undefined at this point..

File: modules/saml/www/sp/saml1-acs.php [hunk 7]
```diff
@@ -66,7 +68,7 @@
     $responseXML = base64_decode($responseXML);
     $isValidated = false; /* Must check signature on response. */
 } else {
-    assert(false);
```
Comment (tvdijen) [inline modules/saml/www/sp/saml1-acs.php]: This leads to undefined variables below.. If we can't handle a response because something is missing, we should just raise an exception.

File: modules/saml/www/sp/saml2-acs.php [hunk 8]
```diff
@@ -51,14 +53,17 @@
     }
 }
 
-$idp = $issuer;
```
Comment (tvdijen) [inline modules/saml/www/sp/saml2-acs.php]: This means we have to test both $idp and $issuer for all possible types.. Therefore I've renamed all instances of `$idp` to `$issuer`

File: modules/saml/www/sp/saml2-acs.php [hunk 9]
```diff
@@ -136,15 +141,16 @@
     // the status of the response wasn't "success"
     $e = $e->toException();
     \SimpleSAML\Auth\State::throwException($state, $e);
+    return;
```
Comment (jaimeperez) [inline modules/saml/www/sp/saml2-acs.php]: This should never be necessary. I'd suggest that instead of this, we fix `throwException()` so that even if the exception handler function returns (which it shouldn't), the method doesn't return itself. Maybe raise an exception warning about a bogus exception handler?

File: modules/saml/www/sp/saml2-acs.php [hunk 10]
```diff
@@ -186,6 +191,7 @@
         $foundAuthnStatement = true;
     }
 }
+$assertion = end($assertions);
```
Comment (jaimeperez) [inline modules/saml/www/sp/saml2-acs.php]: Why is this necessary? Is psalm complaining that `$assertion` is no longer defined outside of the loop's scope?

File: modules/saml/www/sp/saml2-logout.php [hunk 11]
```diff
@@ -12,11 +12,11 @@
 
 $sourceId = substr($_SERVER['PATH_INFO'], 1);
 
+/** @var \SimpleSAML\Module\saml\Auth\Source\SP $source */
 $source = \SimpleSAML\Auth\Source::getById($sourceId);
 if ($source === null) {
     throw new \Exception('Could not find authentication source with id '.$sourceId);
-}
-if (!($source instanceof \SimpleSAML\Module\saml\Auth\Source\SP)) {
+} elseif (!($source instanceof \SimpleSAML\Module\saml\Auth\Source\SP)) {
```
Comment (jaimeperez) [inline modules/saml/www/sp/saml2-logout.php]: The same as with state arrays happen here. If it is actually possible that the auth source is `null`, then we should check for that first, then assert that it's an object of that class in the phpdoc.

File: lib/SimpleSAML/Logger.php [hunk 12]
```diff
@@ -389,9 +389,6 @@ private static function defer($level, $message, $stats)
      */
     private static function createLoggingHandler($handler = null)
     {
-        // set to false to indicate that it is being initialized
-        self::$loggingHandler = false;
```
[the remark this answers is not in the document]
Comment (tvdijen) [reply lib/SimpleSAML/Logger.php]: Hmm, I'm not entirely sure and I even removed more related to this in batch #3:
https://github.com/simplesamlphp/simplesamlphp/pull/1162/files#diff-a7df7b8ccd1feac576070561490e17fc

I think I may have misunderstood what it does.. I can undo this and see what the actual error was

File: modules/core/lib/Auth/Process/AttributeAlter.php (hunk 2 above)
Comment (tvdijen) [reply modules/core/lib/Auth/Process/AttributeAlter.php]: Not currently.. A future (I guess 3.x) version of Psalm will support this even without annotation:
https://psalm.dev/r/41c55a6681
Let me find out the exact version and add it to the annotation so it can be removed in the future.

File: modules/core/lib/Auth/Process/AttributeMap.php [hunk 13]
```diff
@@ -89,7 +89,8 @@ private function loadMapFile($fileName)
             }
             $filePath = Module::getModuleDir($m[0]).'/attributemap/'.$m[1].'.php';
         } else {
-            $filePath = $config->getPathValue('attributenamemapdir', 'attributemap/').$fileName.'.php';
+            $attributenamemapdir = $config->getPathValue('attributenamemapdir', 'attributemap/') ?: 'attributemap/';
+            $filePath = $attributenamemapdir.$fileName.'.php';
```
[the remark this answers is not in the document]
Comment (tvdijen) [reply modules/core/lib/Auth/Process/AttributeMap.php]: `getPathValue()` is broken imho.. It will produce different output for a config-setting set to `null` and a config-setting that is not set at all..

File: modules/core/www/no_cookie.php [hunk 14]
```diff
@@ -11,8 +11,15 @@
...
+/** @var string $header */
+$header = $translator->t('{core:no_cookie:header}');
+/** @var string $desc */
+$desc = $translator->t('{core:no_cookie:description}');
+/** @var string $retry */
+$retry = $translator->t('{core:no_cookie:retry}');
+
+$t->data['header'] = htmlspecialchars($header);
+$t->data['description'] = htmlspecialchars($desc);
+$t->data['retry'] = htmlspecialchars($retry);
```
[the remark this answers is not in the document]
Comment (tvdijen) [reply modules/core/www/no_cookie.php]: I'm in favor of that, however we the phpdoc in Template::t tells says it will not be removed until 2.0...
Let's leave this for now.. Removing the deprecated behaviour will be another PR anyway

File: modules/saml/lib/Auth/Source/SP.php [hunk 15]
```diff
@@ -549,7 +551,7 @@ private function startSSO2(Configuration $idpMetadata, array $state)
         }
 
         if (isset($state['saml:NameID'])) {
-            if (!is_array($state['saml:NameID']) && !is_a($state['saml:NameID'], NameID::class)) {
+            if (!is_array($state['saml:NameID'])) {
```
[the remark this answers is not in the document]
Comment (tvdijen) [reply modules/saml/lib/Auth/Source/SP.php]: I believe it was a RedundantCondition... The state-array doesn't contain objects, or does it?

File: modules/saml/www/sp/discoresp.php [hunk 16]
```diff
@@ -11,7 +11,12 @@
 if (!array_key_exists('idpentityid', $_REQUEST)) {
     throw new \SimpleSAML\Error\BadRequest('Missing idpentityid to discovery service response handler');
 }
+
+/** @var array $state */
 $state = \SimpleSAML\Auth\State::loadState($_REQUEST['AuthID'], 'saml:sp:sso');
+if ($state === null) {
+    throw new \SimpleSAML\Error\NoState();
+}
```
[the remark this answers is not in the document]
Comment (tvdijen) [reply modules/saml/www/sp/discoresp.php]: The annotation should probably go...

File: modules/saml/www/idp/certs.php [hunk 17]
```diff
@@ -18,19 +18,22 @@
...
         $certInfo = SimpleSAML\Utils\Crypto::loadPublicKey($idpmeta, true, 'https.');
         break;
     default:
         throw new \SimpleSAML\Error\NotFound('Unknown certificate.');
 }
-
 header('Content-Disposition: attachment; filename='.substr($_SERVER['PATH_INFO'], 1));
 header('Content-Type: application/x-x509-ca-cert');
+
 echo $certInfo['PEM'];
```
[the remark this answers is not in the document]
Comment (tvdijen) [reply modules/saml/www/idp/certs.php]: Nope, annotations only work on assignments.

File: modules/saml/www/proxy/invalid_session.php [hunk 18]
```diff
@@ -16,10 +16,18 @@
...
+    if ($state === null) {
+        throw new \SimpleSAML\Error\NoState();
+    }
 } catch (\Exception $e) {
     // the user probably hit the back button after starting the logout, try to recover the state with another stage
+    /** @var array $state */
     $state = \SimpleSAML\Auth\State::loadState($_REQUEST['AuthState'], 'core:Logout:afterbridge');
+    if ($state === null) {
+        throw new \SimpleSAML\Error\NoState();
+    }
```
[the remark this answers is not in the document]
Comment (tvdijen) [reply modules/saml/www/proxy/invalid_session.php]: Can probably do without the annotation

File: modules/saml/www/sp/saml2-logout.php (hunk 11 above)
Comment (tvdijen) [reply modules/saml/www/sp/saml2-logout.php]: Unfortunately annotations only work on assignments.. We would have to use assert() to go with your approach

File: modules/core/lib/Auth/Process/AttributeAlter.php (hunk 2 above)
Comment (jaimeperez) [reply modules/core/lib/Auth/Process/AttributeAlter.php]: Great!

File: modules/saml/www/proxy/invalid_session.php (hunk 18 above)
[the remark this answers is not in the document]
Comment (jaimeperez) [reply modules/saml/www/proxy/invalid_session.php]: That's true, psalm should be smart enough to notice that...

File: modules/core/lib/Auth/Process/AttributeMap.php (hunk 13 above)
[the remark this answers is not in the document]
Comment (jaimeperez) [reply modules/core/lib/Auth/Process/AttributeMap.php]: Yeah, that's unfortunate. We should probably fix that behaviour, but it's also kind of self-inflicted pain, since the default value can be of any type. In any case, fixing that would be a bit tricky and require more work to make sure nothing breaks, I think...

File: modules/saml/www/sp/saml2-logout.php (hunk 11 above)
Comment (jaimeperez) [reply modules/saml/www/sp/saml2-logout.php]: Uhm, I was pretty sure I tried using annotations at random places in the code and that worked fine, but I might be wrong... 🤔

File: modules/core/www/no_cookie.php (hunk 14 above)
[the remark this answers is not in the document]
Comment (jaimeperez) [reply modules/core/www/no_cookie.php]: Yep, but I was thinking about removing the documentation for the old usage, not the actual implementation for it. In any case, it's probably better to leave that out for now...

File: modules/saml/lib/Auth/Source/SP.php (hunk 15 above)
[the remark this answers is not in the document]
Comment (jaimeperez) [reply modules/saml/lib/Auth/Source/SP.php]: I believe it does. E.g. metadata objects...

File: modules/core/lib/Auth/Process/AttributeMap.php (hunk 13 above)
[the remark this answers is not in the document]
Comment (tvdijen) [reply modules/core/lib/Auth/Process/AttributeMap.php]: Agreed, so that's why I went with this solution.. Pretty? No! Works? Yes!

File: modules/saml/www/sp/saml2-logout.php (hunk 11 above)
Comment (tvdijen) [reply modules/saml/www/sp/saml2-logout.php]: He're proof: https://psalm.dev/r/262a290058

Comment (jaimeperez) [reply modules/saml/www/sp/saml2-logout.php]: Interesting! Also for the interactive tool that I didn't know!

Comment (tvdijen) [reply modules/saml/www/sp/saml2-logout.php]: Yeah, it's great! I think it's fairly new..
They will ask you to use it if you report a bug so that they can rule out that the bug still exists in the latest version and so they have means of reproducing a bug... Really nice!

File: modules/admin/lib/ConfigController.php [hunk 19]
```diff
@@ -385,7 +385,7 @@ protected function getWarnings()
                     curl_setopt($ch, CURLOPT_PROXYUSERPWD, $this->config->getValue('proxy.auth', null));
                     $response = curl_exec($ch);
 
-                    if (curl_getinfo($ch, CURLINFO_HTTP_CODE) === 200) {
+                    if (!is_bool($response) && (curl_getinfo($ch, CURLINFO_HTTP_CODE) === 200)) {
                         $latest = json_decode($response, true);
```
[the remark this answers is not in the document]
Comment (tvdijen) [reply modules/admin/lib/ConfigController.php]: Note to self:
CURLINFO_HTTP_CODE is a legacy alias of CURLINFO_RESPONSE_CODE.
Should change while at it

File: modules/core/lib/Auth/Process/AttributeAlter.php (hunk 2 above)
Comment (tvdijen) [reply modules/core/lib/Auth/Process/AttributeAlter.php]: I gave up on this one, restored it, and just suppressed the issue..

File: modules/core/lib/Storage/SQLPermanentStorage.php [hunk 20]
```diff
@@ -32,7 +32,7 @@ public function __construct($name, $config = null)
             $config = Configuration::getInstance();
         }
 
-        $datadir = $config->getPathValue('datadir', 'data/');
+        $datadir = $config->getPathValue('datadir', 'data/') ?: 'data/';
```
[the remark this answers is not in the document]
Comment (tvdijen) [reply modules/core/lib/Storage/SQLPermanentStorage.php]: As we agreed above;  no easy way to fix this without breaking stuff. marking as resolved

File: modules/core/www/frontpage_config.php [hunk 21]
```diff
@@ -75,8 +75,8 @@
             curl_setopt($ch, CURLOPT_PROXYUSERPWD, $config->getValue('proxy.auth', null));
             $response = curl_exec($ch);
 
-            if (curl_getinfo($ch, CURLINFO_HTTP_CODE) === 200) {
-                $latest = json_decode($response, true);
+            if (!is_bool($response) && (curl_getinfo($ch, CURLINFO_HTTP_CODE) === 200)) {
+                $latest = json_decode(strval($response), true);
```
[the remark this answers is not in the document]
Comment (tvdijen) [reply modules/core/www/frontpage_config.php]: Was resolved

File: lib/SimpleSAML/Logger.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (tvdijen) [reply lib/SimpleSAML/Logger.php]: Reverted; marking as resolved

File: modules/saml/lib/Auth/Source/SP.php (hunk 15 above)
[the remark this answers is not in the document]
Comment (tvdijen) [reply modules/saml/lib/Auth/Source/SP.php]: Metadata as object? Isn't metadata always an array?
Anyway, I've reverted the change.. Let's see what the actual issue was

File: modules/saml/lib/Auth/Source/SP.php [hunk 22]
```diff
@@ -829,8 +833,10 @@ public function reauthenticate(array &$state)
 
         $session = Session::getSessionFromRequest();
         $data = $session->getAuthState($this->authId);
-        foreach ($data as $k => $v) {
-            $state[$k] = $v;
+        if ($data !== null) {
+            foreach ($data as $k => $v) {
+                $state[$k] = $v;
+            }
```
[the remark this answers is not in the document]
Comment (tvdijen) [reply modules/saml/lib/Auth/Source/SP.php]: It's part of the pubic SP API and can be called at any point in time...
I'm not sure what to do with this...

File: modules/saml/www/proxy/invalid_session.php (hunk 18 above)
[the remark this answers is not in the document]
Comment (tvdijen) [reply modules/saml/www/proxy/invalid_session.php]: Annotations removed

File: modules/saml/www/sp/saml2-acs.php (hunk 9 above)
Comment (tvdijen) [reply modules/saml/www/sp/saml2-acs.php]: Done

File: modules/saml/www/sp/saml2-acs.php (hunk 10 above)
Comment (tvdijen) [reply modules/saml/www/sp/saml2-acs.php]: Exactly; if `$assertions` would be empty, `$assertion` would never be defined.

File: modules/saml/lib/Auth/Source/SP.php (hunk 22 above)
[the remark this answers is not in the document]
Comment (jaimeperez) [reply modules/saml/lib/Auth/Source/SP.php]: Uhm, that's unfortunate. Then I'd suggest to invert the check and if `$data` is `null`, throw an exception.

File: modules/saml/www/proxy/invalid_session.php (hunk 18 above)
[the remark this answers is not in the document]
Comment (jaimeperez) [reply modules/saml/www/proxy/invalid_session.php]: Actually, looking into `loadState()`, since we're not passing `true` to the third argument (`$allowMissing`), if we cannot find the state, we'll either redirect or thrown an exception, meaning `$state` can never be `null` and then these checks are superfluous...

[the remark this answers is not in the document]
Comment (tvdijen) [reply modules/saml/www/proxy/invalid_session.php]: Oh good catch! I totally missed that!

File: modules/saml/lib/Auth/Source/SP.php (hunk 22 above)
[the remark this answers is not in the document]
Comment (tvdijen) [reply modules/saml/lib/Auth/Source/SP.php]: Done = resolveed

File: modules/saml/lib/Auth/Source/SP.php (hunk 15 above)
[the remark this answers is not in the document]
Comment (tvdijen) [reply modules/saml/lib/Auth/Source/SP.php]: The issue was actually a bug introduced in #1008.. The check for NameID objects was added, but the rest of the code was never prepared for objects instead of arrays.. Resolved!
